### ISSUE TITLE
High CPU Usage by StackExchange dll

### ISSUE THREAD

**Issue by heera-jaiswal** [StackExchange/StackExchange.Redis#654]
Hi,

I notice very high CPU usage(80-90 %) of the app server when I give concurrent requests using 100 threads to the server which is using stackExchange lib to access redis server. 
I used debug diagnosis tool to figure out the issue. It seems StackExchange is taking most of the CPU
![image](https://user-images.githubusercontent.com/4180812/27687594-0b60c4ee-5cf5-11e7-8caf-0c29cd9b2b28.png)

This is how I'm initializing ConnectionMultiplexer

Here is the code of same
`

public class SE_Redis{
	
		private static ConnectionMultiplexer masterClientPool = null;
		public static IDatabase getClient()
        {
            if (masterClientPool == null || !masterClientPool.IsConnected)
            {                                
                masterClientPool = initMultiplexer("SE_REDIS_GDS_MASTER");
            }
            
            if (multiplexer != null)
            {                
                return multiplexer.GetDatabase();
            }
            else
            {
                throw new Exception("Failed to init redis");
            }
                
        }
        private static ConnectionMultiplexer initMultiplexer(string settingkey)
        {
            
            string connString = ConfigurationManager.AppSettings[settingkey].ToString();            
            string clientName = getClientName();            
            ConfigurationOptions options = ConfigurationOptions.Parse(connString);
            options.ClientName = clientName;            
            var multiplexer = ConnectionMultiplexer.Connect(options);            
            return multiplexer;
        }
}
        public CompanyFields getCompanyDetails(int provId,string provCompId)
        {
            string[] eVars = new string[] { provId.ToString(), provCompId };
            string redisKey = this.getRedisKey(eVars);
            string result = null;
            var client = SE_Redis.getClient();            
            result = client.StringGet(redisKey);           
            if (string.IsNullOrEmpty(result))
            {
                return null;
            }
            else
            {
                return Jil.JSON.Deserialize<CompanyFields>(result);
            }
        }
`
Please help if I'm missing anything in initialization

**Comment by heera-jaiswal** [comment]
Other clients like [redis-py](https://redis-py.readthedocs.io/en/latest/) works fine on concurrent requests. I'm facing this issue only on c# client.

Please help

**Comment by heera-jaiswal** [comment]
Result from newrelic Thread profiling for CPU usage
![image](https://user-images.githubusercontent.com/4180812/27731606-5591ec9e-5dab-11e7-9c9e-2a412da7a1eb.png)

**Comment by NickCraver** [comment]
This is one of those cases where a profiler is not showing what you think it is. Profilers don't really know how much CPU a thread is using, they *estimate* how much CPU a thread is using and the way they do it is tremendously inaccurate for idle things.

What happens is a profiler takes snapshots of what's "in progress" (what are the current stack traces?) many times. Let's say every 100ms for 10 seconds. Now for those 100 snapshots, each one looks something like this:

```
WriteAllQueues(); // stack
SometMethod1(): // stack
SometMethod2(): // stack
```
```
WriteAllQueues(); // stack
SometMethod3(): // stack
SometMethod4(): // stack
```
...etc.

Now the problem here is `WriteAllQueues()` is a dedicated thread, so it's *always there*, **even when idle**. By "wall time", it's there, but it's **mostly idle**. But since the profiler uses these snapshots to estimate "CPU time", it's wildly inaccurate. You can reproduce this by simple starting a thread, doing nothing, and taking a profile...it will show as 100% CPU cost or 1/n cost even though it's actually doing nothing. You can make it `Thread.Sleep(1000000)` to see it.

It's possible there's actual work being done here, but given that 32% of you time in that shot drops to 0 for any child method, I'd say the situation above is exactly what's happening. I agree it's confusing, but if you realize what the profiler is doing, it starts to make sense.

Now for the rest - it looks like there's a lot of timers and thread pool action happening. Do you have something async or I/O happening on every request?

**Comment by heera-jaiswal** [comment]
Thanks for reply.

I understood your point. But I should have given more info on issue.
I wanted to test load on api server which is using stack exchange.

Here is some data from a test api client, 
server config: Aws Windows server with 4 virtual processor.
1)
Total requests: 5000
Parallel requests: 1
Avg response time: 35 ms

2)
Total requests: 5000
Parallel requests: 2
Avg response time: 80 ms

3)
Total requests: 5000
Parallel requests: 10
Avg response time: 165 ms

4)
Total requests: 5000
Parallel requests: 100
Avg response time: 701 ms

When case 4) was running, I checked server's cpu was continuously high (80-100%). But it was fine in case of 1). So I profiled the iis worker process and result was the same I have sent you earlier.

And there is no other disk or Network IO involved apart from Redis. Neither any Async code was executing.
Peak redis throughput during case 4) was 75k calls/Minute

So 701 ms would be expected response time on such config with the given load ?

**Comment by alxbog** [comment]
Have the same problem. The more concurrent requests are made the more times it takes for each one on average.

I've created a sample console app that makes 100000 requests in a short amount of time against a single `ConnectionMultiplexer`. A single `StringSet` operation takes 70 secs (!) on average. CPU consumption is nearly 100%.

That's interesting is that when I run exactly the same code in LinqPad it runs very fast (lest than 200 ms per request).

**Comment by alxbog** [comment]
@NickCraver Please take a look at this problem. It's very serious issue.

We faced this problem on our production deployment, not in an artificial sample I've created. This issue was very difficult to track (no exceptions, just slow performance), I spent several days to reduce it to the sample I've attached in the previous message.

**Comment by alxbog** [comment]
I found that the difference was in attached VS debugger. It becomes really slow with debugger attached. But we also have the same slow performance behaviour on our Azure worker role (Cloud Services) production deployment.

**Comment by alxbog** [comment]
We solved the problem by using faster VM for our app on Azure.
Still this behaviour seems odd. VM CPU load was less than 80% before upgrade and everything else except Redis client lib worked just fine.

**Comment by NickCraver** [comment]
It wasn't apparent it overlapped, but it is not in hindsight. If anyone still seeing this after upgrading to 1.2.6 or higher? There was a nasty completion thread bug in the build that we recognized and fixed in the 1.2.6 build.

**Comment by mgravell** [comment]
Note; 2.0 completely changes the network core: https://github.com/StackExchange/StackExchange.Redis/issues/871 - probably makes sense to close this as it will change this completely
